### PR TITLE
3771 mark post as reply

### LINKED ISSUE #3771: Show that a post is part of a thread
Hi! Thanks for creating Tusky!

I'd love it if toots in the timeline showed that they're part of a thread. Currently they are presented in the feed in the same way as standalone posts, and you don't know that you're missing the context.

This is how it looks in the Megalodon app:

timeline in the Megalodon app

* * * *
- Tusky Version: 22.0
- Android Version: 11
- Android Device: N/A
- Mastodon instance (if applicable): N/A

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by Lakoja**
Fixes #3771 / partly addresses #1469

grafik

This shows in a timeline the status info "In reply to < username>" just like the existing "< username> boosted" above a status.

This is a more or less low-hanging fruit as the information is available and the web gui does this (but see review comments here for details).

This is NOT the requested "Show thread" functionality - which would still need to be discussed; i.e. every reply makes a thread?

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt [hunk 1]
```diff
@@ -111,6 +111,8 @@ data class TimelineStatusWithAccount(
     val status: TimelineStatusEntity,
     @Embedded(prefix = "a_")
     val account: TimelineAccountEntity? = null, // null when placeholder
-    @Embedded(prefix = "rb_")
+    @Embedded(prefix = "rp_")
```
Comment (Lakoja) [inline app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt:114]: Note: The source of this account data is a bit unexpected.

Most other information comes directly from the Api. This here is mixed-in only locally. This could be surprising?

File: app/src/main/res/drawable/ic_reply_24dp.xml [hunk 2]
```diff
@@ -5,6 +5,6 @@
     android:viewportHeight="24.0"
     android:viewportWidth="24.0">
     <path
-        android:fillColor="#fff"
+        android:fillColor="?android:attr/textColorTertiary"
```
[the remark this answers is not in the document]
Comment (nikclayton) [reply app/src/main/res/drawable/ic_reply_24dp.xml:8]: I think the correct thing to do is to leave `fillColor` as is (or, better, use `@android:color/white`), and set `android:tint` on the `vector`. See `ic_edit_24dp.xml` for an example.

[the remark this answers is not in the document]
Comment (Lakoja) [reply app/src/main/res/drawable/ic_reply_24dp.xml]: Ok.
(Changed it back to fff, which is the most consistent with other icons - not all do this, though.)

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt [hunk 3]
```diff
@@ -29,9 +33,13 @@ import retrofit2.HttpException
 @OptIn(ExperimentalPagingApi::class)
 class NetworkTimelineRemoteMediator(
     private val accountManager: AccountManager,
-    private val viewModel: NetworkTimelineViewModel
+    private val viewModel: NetworkTimelineViewModel,
+    db: AppDatabase,
+    private val gson: Gson
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:38]: I think all of these can be injected here.

Then you can inject the remote mediator in to the view model. That way the view model doesn't need to receive `db` or `gson` as parameters.

File: app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt (hunk 1 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt:114]: "rp_" is "reply"?

What's "bo_"? "boost"?

That's probably sufficiently confusing that it should just be "reply_" and "reblog_".

I prefer "reblog" to "boost" because we're pretty consistent about using the API terms for these things in the code, not the UI terms (e.g., it's always "status" not "post", etc).

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -80,6 +80,7 @@
 
     <string name="post_username_format">\@%s</string>
     <string name="post_boosted_format">%s boosted</string>
+    <string name="post_replied_format">In reply to %s</string>
```
Comment (nikclayton) [inline app/src/main/res/values/strings.xml:83]:
```suggestion
    <string name="post_replied_format">%s replied</string>
```

?

Or is the %s here the name of the original author, and not the person who replied?

**Review by nikclayton — CHANGES_REQUESTED** [review]
A couple of low-hanging things I spotted. Later today I'll check this out locally and take a proper look.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (Lakoja) [reply app/src/main/res/values/strings.xml:83]: The person who replied is the status owner (and thus listed one line below).

%s is the "replied to account name".

File: app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt (hunk 1 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/db/TimelineStatusEntity.kt]: Ok, changed it to "author", "replied" and "reblogger" (all 3 TimelineStatusEntity).

(Generally these 2 queries and tedious to change and to understand.)

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt (hunk 3 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:38]: The remote mediator has the view model as dependency

(and thus a dependency cylce)

**Comment by nikclayton** [comment]
I haven't tested this yet, but I've just observed the following:

1. `item_status.xml` (and related layouts) have a `R.id.status_info` view at the top, which is where the "X boosted" message goes.
2. `StatusBaseViewHolder.java` has a `setIsReply` method that's called whenever the status is a reply

So, in `StatusViewHolder.java`, wouldn't it be simpler to just do this:

```java
protected void setIsReply(boolean isReply) {
    super setIsReply(isReply);
    if (isReply) {
        statusInfo.text = getString(R.string.post_replied_format, ...)
    }
}
```

?

Also, what should happen if the status was a reply, and it was reblogged? https://chaos.social/@jollysea/110668273086228868 is an example post that is a reply that was reblogged by @mcclure.

1. The "X boosted" message is more important
2. The "In reply to Y" message is more important
3. The message should be "X boosted this reply to Y"
4. Show two lines of text

Note that that status above is an interesting edge case for this, because it's mcc reblogging a reply to her own post, so although option 3 might be generally good, it would lead to "mcc boosted this reply to mcc". Which might be more mcc than a single line can handle :-)

**Comment by Lakoja** [comment]
> So, in `StatusViewHolder.java`, wouldn't it be simpler to just do this:
> 
> ```java
> protected void setIsReply(boolean isReply) {
>     super setIsReply(isReply);
>     if (isReply) {
>         statusInfo.text = getString(R.string.post_replied_format, ...)
>     }
> }
> ```
> 
> ?

Instead of what?
At least all the changes before and in (now) `setStatusInfoText()` would be necessary as they also deal with emojis and a maybe different name.
And at least at the moment the `setIsReply()` only deals with two different reply cases. The reblog information is orthogonal.

**Comment by Lakoja** [comment]
> 
> Also, what should happen if the status was a reply, and it was reblogged? https://chaos.social/@jollysea/110668273086228868 is an example post that is a reply that was reblogged by @mcclure.
> 
>     3. The message should be "X boosted this reply to Y"
> 

I kind of like option 3.
However that might be too much information? (and will often be too long, names tend to be long...)
2 lines will also be too much, I think.

At the moment (in my code) reblog is more important than reply. So an "if ... else if" logic.

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt [hunk 5]
```diff
@@ -71,7 +80,13 @@ class NetworkTimelineRemoteMediator(
                 val expanded = oldStatus?.isExpanded ?: activeAccount.alwaysOpenSpoiler
                 val contentCollapsed = oldStatus?.isCollapsed ?: true
 
+                var inReplyToAccount:TimelineAccount? = null
+                if (status.inReplyToAccountId != null) {
+                    inReplyToAccount = accountDao.get(status.inReplyToAccountId)?.toAccount(gson)
```
[the remark this answers is not in the document]
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:84]: I don't think the account will always exist in the local database.

Consider: You follow user B, they reply to user A, who you do not follow.

This works for a reblog, because the `reblog` field contains the complete status, which includes the full information of the original author (https://docs.joinmastodon.org/entities/Status/#reblog).

A status that's a reply doesn't, it only contains `in_reply_to_account_id` (https://docs.joinmastodon.org/entities/Status/#in_reply_to_account_id).

It *might* also contain a `mentions` object, but AFAIK there's no requirement that a reply also mention the person they're replying to, and even if there is a `mentions` object it doesn't contain the display name of the person that's being replied to.

[the remark this answers is not in the document]
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:84]: It does work so far (for a week) on my timeline.

Not sure if the api actually provides arbitrary account information but I'm pretty sure we wouldn't want to do an api call for this information.

At the moment if an account is missing the text line would be missing.

(So is there an alternative?)

[the remark this answers is not in the document]
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:84]: Two strings, "Replied" and "Replied to %s", select the correct one based on whether or not we know the account info.

File: app/src/main/res/layout/item_status.xml [hunk 6]
```diff
@@ -17,7 +17,7 @@
         android:layout_marginStart="14dp"
         android:layout_marginTop="@dimen/status_reblogged_bar_padding_top"
         android:layout_marginEnd="14dp"
-        android:drawableStart="@drawable/ic_reblog_18dp"
+        android:drawableStart="@drawable/ic_reblog_24dp"
```
Comment (nikclayton) [inline app/src/main/res/layout/item_status.xml:20]: Why is this switching to the bigger icon? It causes the effective left margin of the text to change.

Comment (nikclayton) [reply app/src/main/res/layout/item_status.xml:20]: Also, you could copy `ic_reply_24dp.xml` to ic_reply_18dp.xml`, and then use that for replies.

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineRemoteMediator.kt:84]: ok

**Comment by nikclayton** [comment]
@Lakoja Can you take a look at https://github.com/Lakoja/Tusky/pull/4, and see if you want to merge that in to this PR? I figured that was easier than more back-and-forth.

You'll want to merge `develop` in to this PR first to minimse the diffs.

Some quick screenshots:

## It shows reply and boost, with a bullet between them

image

image

--- 

## If the account isn't known, but mentioned, it uses the @-form of the user name

image

---

## Long lines are truncated / ellipsised

image

---

I'm not 100% convinced that moving the left margin over is a good idea, but as you can see from the screenshots, sometimes the line does get quite long.

WDYT?

File: app/src/main/res/layout/item_status.xml (hunk 6 above)
Comment (Lakoja) [reply app/src/main/res/layout/item_status.xml:20]: The size is determined by the component (height of TextView).

Already before: The text behind the drawable is not alligned with the name or the content below. This does not change here.

So I could leave it as 18 but that would not change it graphically.

**Comment by Lakoja** [comment]
> Some quick screenshots:

> WDYT?

I find the lines with two information quite hard to parse visually.
(thus) I still wouldn't present two informations there.

**Comment by nikclayton** [comment]
> > Some quick screenshots:
> 
> > WDYT?
> 
> I find the lines with two information quite hard to parse visually. (thus) I still wouldn't present two informations there.

I think that might be a problem, because with this PR we're trying to use the statusInfo to show two difference pieces of information.

Currently the statusInfo answers the "Why am I seeing this in my timeline?" question. It tells the user "You are seeing this because someone you follow has boosted it".

This PR extends that to also answer the "Can I tap on this post to read a thread?" question.

I'm not sure that only ever showing one bit of information is a good idea, because it means that users will not be able to rely on the information in the statusInfo -- what's shown will vary, and it will be difficult for users to figure out why it's varying.

By which I mean that some replies -- replies from people that the user follows -- will be marked as replies, and the user will know they can tap the post to see the thread.

But other replies -- replies from people they don't follow that have been boosted by someone they do -- will not be marked as replies.

[I know the "reply" icon at the bottom of the post will change as well, but as we see from the bug reports, this is also something that user's don't always notice]

If we flip that example around, and decide that showing that it's part of a thread is more important than showing that it's a boost, then we get the problem that a user can see "user-i-dont-follow replied to ..." in the statusInfo, and not know why this post is appearing in their timeline.

The first screenshot in my previous comment is an example of that. I don't follow rain, but I do follow Norm.

Both of those seem like a recipe for user confusion.

I think we can collapse some of the information down. For example, if someone boosts their own post there's no need to show their  display name again.

And we could experiment with putting some of the information elsewhere. I'm not sure a two-line statusInfo is a good idea, but maybe the "reply" icon could appear next to the display name of the poster (i.e., the displayname that appears to the right of the avatar)? That would leave the statusInfo for boosts (and later, hashtags).

But... that would mean the left-edge of the display names would jump around a bit as the user scrolls through their timeline.

I'll try some mockups and see what happens.

**Comment by nikclayton** [comment]
Here's the latest variant (I'm still waiting for a self-boost to cross my timeline).

## Replying to themselves

image

**Comment by nikclayton** [comment]
Some of https://github.com/tuskyapp/Tusky/issues/3883 is relevant to this as well. There's one, possibly two additional values that can be included in the status_info, but they're mutually exclusive with the others (I think). They are:

- "Sent you a direct message"
- "Replied to your direct message"

or similar. Possibly with "direct message" in that text in bold, or bold tusky_red, or something like that, to make it even more distinctive.

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt [hunk 7]
```diff
@@ -207,8 +207,8 @@ fun TimelineStatusWithAccount.toViewData(gson: Gson, isDetailed: Boolean = false
             id = status.serverId,
             url = null, // no url for reblogs
             account = this.reblogAccount!!.toAccount(gson),
-            inReplyToId = null,
-            inReplyToAccountId = null,
+            inReplyToId = status.inReplyToId,
```
[the remark this answers is not in the document]
Comment (Tak) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt:210]: But the boost itself is not the reply, right? The *actionable* status is the reply

File: app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java [hunk 8]
```diff
@@ -65,19 +67,34 @@ public void setupWithStatus(@NonNull StatusViewData.Concrete status,
...
         if (payloads == null) {
-
             boolean sensitive = !TextUtils.isEmpty(status.getActionable().getSpoilerText());
             boolean expanded = status.isExpanded();
 
             setupCollapsedState(sensitive, expanded, status, listener);
 
             Status reblogging = status.getRebloggingStatus();
-            if (reblogging == null || status.getFilterAction() == Filter.Action.WARN) {
+            boolean isReply = status.getStatus().getInReplyToId() != null;
```
Comment (Tak) [inline app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java:76]: Don't you need to use `status.getActionable` here, and on line 90?

**Review by Tak — COMMENTED** [review]
Doesn't this need a database migration for `TimelineAccountEntity`?

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt:210]: What is boosted ("status") can be a reply (too).

This info is "unused".

**Comment by Lakoja** [comment]
> Doesn't this need a database migration for `TimelineAccountEntity`?

There is no change there?

All info here is present (in rare cases the "replied to account info" might be missing and then the name is not show).

File: app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java (hunk 8 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java:76]: I was searching for some time for the right location with the right information.
"actionable" was not it if I remember correctly.

It is working and plausible at the moment.

(Would need to look deep(er) into the code if you want me to be 100% certain about this.)

**Comment by mcclure** [comment]
What does this patch do if a post is BOTH a reply AND an RT?

File: app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tak) [reply app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineTypeMappers.kt:210]: Ah, I see now, this is here as a hack to populate the inreplyto timeline account of the concrete statusviewdata when it's a boost of a reply.
…but as you say, we're not displaying that information, so it could be reverted to avoid confusing the issue (although the status/actionable thing is already confusing)

File: app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java (hunk 8 above)
Comment (Tak) [reply app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java:76]: You're right, the inreplyto timeline account is only available on status, but without the timeline type mapper change, the correct thing to check for the inreplyto id is the actionable status. (And if `status.getActionable().getInReplyToId()` is nonnull, then `status.getInReplyToAccount()` will be as well.)

**Comment by Tak** [comment]
The current PR code displays this when a reply is boosted
Screenshot 2023-10-06 at 11 12 35 AM

**Comment by Tak** [comment]
Oh, it looks super weird in the thread view - is that intentional?

**Comment by Tak** [comment]
I guess this change should be applied to the notification timeline as well? (Maybe in a second pass after the refactor lands?)

**Comment by Lakoja** [comment]
> Oh, it looks super weird in the thread view - is that intentional?

No.
(Super weird: it shows "Replied" above every status.)

I fixed it by not showing the status info in the thread view.

But I guess the introduced flag should/could be part of "StatusDisplayOptions" object?

**Comment by Lakoja** [comment]
> I guess this change should be applied to the notification timeline as well? (Maybe in a second pass after the refactor lands?)

Yes, probably and yes.

File: gradle/libs.versions.toml [hunk 9]
```diff
@@ -120,6 +120,7 @@ glide-compiler = { module = "com.github.bumptech.glide:ksp", version.ref = "glid
 glide-core = { module = "com.github.bumptech.glide:glide", version.ref = "glide" }
 glide-okhttp3-integration = { module = "com.github.bumptech.glide:okhttp3-integration", version.ref = "glide" }
 gson = { module = "com.google.code.gson:gson", version.ref = "gson" }
+kapt = { module = "androidx.room:room-compiler", version.ref = "androidx-room" }
```
Comment (Lakoja) [inline gradle/libs.versions.toml:123]: Not sure why this is needed _now_ (starting today)?

If I don't have this there will be a build error that the (automatically generated) AppDatabase_Impl does not implement all needed dao factory methods.

File: app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java (hunk 8 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java:77]: What is the meaning of "actionable" in current code?

I see that it only has a value if "reblog" is true (Status.kt:59), so the thinking is probably "the item the action was taken for"?

(The "reply stuff" I have aligned more with the extsting "inReplyToAccountId".)

Comment (Tak) [reply app/src/main/java/com/keylesspalace/tusky/adapter/StatusViewHolder.java:77]: Yeah, my understanding is that the "outer" status is the boost itself, and the actionable status is the status that was boosted

**Comment by Tak** [comment]
imo the changes look ok, but it will have to be rebased again after #4026 lands (after the 25.0 release) 😐 
I'm really sorry that this has been in stasis for almost a year

**Comment by connyduck** [comment]
Should we in turn get rid of the double arrow on the reply button that currently indicates replies? It wouldn't be necessary anymore.

File: gradle/libs.versions.toml (hunk 9 above)
Comment (connyduck) [reply gradle/libs.versions.toml:111]: This should not be necessary as we migrated Room to ksp. If you get an error, try a clean build.

Comment (Lakoja) [reply gradle/libs.versions.toml]: Removed.

**Comment by Lakoja** [comment]
Replaced by: https://github.com/tuskyapp/Tusky/pull/4834
